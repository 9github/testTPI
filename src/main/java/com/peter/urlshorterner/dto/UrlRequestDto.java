package com.peter.urlshorterner.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Builder(toBuilder = true)
@NoArgsConstructor
@AllArgsConstructor
public class UrlRequestDto {

  private String fullUrl;

  private String customAlias;
}
