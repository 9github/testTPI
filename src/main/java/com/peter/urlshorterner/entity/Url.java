package com.peter.urlshorterner.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Builder(toBuilder = true)
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = false)
@Entity()
public class Url {

  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  @Column(name = "id", nullable = false, unique = true, length = 36)
  @EqualsAndHashCode.Include
  private String id;

  @Column(name = "alias")
  private String alias;

  @Column(name = "full_url")
  private String fullUrl;

  @Column(name = "short_url")
  private String shortUrl;
}
