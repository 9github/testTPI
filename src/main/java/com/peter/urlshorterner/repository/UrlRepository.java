package com.peter.urlshorterner.repository;

import com.peter.urlshorterner.entity.Url;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
@Transactional(readOnly = true)
public interface UrlRepository extends JpaRepository<Url, String> {

  Url findByLongUrl(String longUrl);

  boolean existsByLongUrl(String longUrl);
  boolean existsByShortUrl(String shortUrl);
}
