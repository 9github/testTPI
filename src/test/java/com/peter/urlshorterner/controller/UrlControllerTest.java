package com.peter.urlshorterner.controller;

import static com.peter.urlshorterner.util.TestUtil.getBodyAsListOf;
import static io.zonky.test.db.AutoConfigureEmbeddedDatabase.DatabaseProvider.ZONKY;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.peter.urlshorterner.dto.UrlDto;
import com.peter.urlshorterner.entity.Url;
import com.peter.urlshorterner.repository.UrlRepository;
import io.zonky.test.db.AutoConfigureEmbeddedDatabase;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

@AutoConfigureMockMvc
@SpringBootTest
@AutoConfigureEmbeddedDatabase(provider = ZONKY)
@ExtendWith({SnapshotExtension.class})
public class UrlControllerTest {

  @Autowired private MockMvc mockMvc;
  @Autowired UrlRepository urlRepository;
  private Expect expect;

  @BeforeEach
  void setUp() {
    Url url = Url.builder().alias("alias").fullUrl("fullUrl").shortUrl("shortUrl").build();
    urlRepository.save(url);
  }

  @AfterEach
  void tearDown() {
    urlRepository.deleteAll();
  }

  final UrlDto urlDto =
      UrlDto.builder().alias("alias").fullUrl("fullUrl").shortUrl("shortUrl").build();

  @Test
  void getUrls() throws Exception {
    var body =
        getBodyAsListOf(
            UrlDto.class,
            mockMvc
                .perform(get("/urls").contentType(MediaType.APPLICATION_JSON))
                .andExpect(status().is(200))
                .andReturn()
                .getResponse()
                .getContentAsString());

    expect.serializer("json").toMatchSnapshot(body);
  }

  @Test
  void getUrlByAlias() throws Exception {
    mockMvc
        .perform(get("/alias").contentType(MediaType.APPLICATION_JSON))
        .andExpect(status().is(302))
        .andExpect(header().string("location", "fullUrl"))
        .andReturn();
  }

  @Test
  void deleteUrlByAlias() throws Exception {
    final String ALIAS = "alias";
    mockMvc
        .perform(delete("/" + ALIAS).contentType(MediaType.APPLICATION_JSON))
        .andExpect(status().is(204))
        .andReturn();

    assertFalse(urlRepository.existsByAlias(ALIAS));
  }
}
