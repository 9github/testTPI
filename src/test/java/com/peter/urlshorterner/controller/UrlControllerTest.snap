com.peter.urlshorterner.controller.UrlControllerTest.getUrls=[
  [
    {
      "alias": "alias",
      "fullUrl": "fullUrl",
      "shortUrl": "shortUrl"
    }
  ]
]